[PASS] InstrumentBoundCheckers pass

denis0x0D opened the pull request:
The patch is related to issue:
https://discuss.tvm.ai/t/array-bounds-checking/944

Lets imagine simple example with out of bounds access:
 
```
n = tvm.var("n")
 
 A = tvm.placeholder ((n,n,n), name='A')
 B = tvm.placeholder ((n,n,n), name='B')
 C = tvm.compute(A.shape, lambda i,j,k: A[i + 300][j][k] + B[i][j][k], name='C')
 s = tvm.create_schedule (C.op)

```

We will get IR like this:
```
 produce C {
    for (i, 0, n) {
      for (j, 0, n) {
        for (k, 0, n) {
          C[((((i*n) + j)*n) + k)] = (A[(((((i + 300)*n) + j)*n) + k)] + B[((((i*n) + j)*n) + k)])
        }
      }
    }
  }
```
This is actually out of the bounds access, this code could cause a segfault or access to "junk" memory since TMV has native runtime.

Could we set for example, in debug mode, an instrumentation before every memory access with simple check and assert.  So at the runtime we can handle it, and make sure our model does not access any invalid memory.

 ```
produce C {
    for (i, 0, n) {
      for (j, 0, n) {
        for (k, 0, n) {
          if (((((((((i + 300)*n) + j)*n) + k) < ((n*n)*n)) && (((((i*n) + j)*n) + k) < ((n*n)*n))) && (((((i*n) + j)*n) + k) < ((n*n)*n)))) {
            C[((((i*n) + j)*n) + k)] = (A[(((((i + 300)*n) + j)*n) + k)] + B[((((i*n) + j)*n) + k)])
          } else {
            assert(((((((((i + 300)*n) + j)*n) + k) < ((n*n)*n)) && (((((i*n) + j)*n) + k) < ((n*n)*n))) && (((((i*n) + j)*n) + k) < ((n*n)*n))), "OUT OF BOUNDS")
            1
          }
        }
      }
    }
  }
```

So, I wrote the pass, which instruments bound check before potentially invalid memory access.

The pass could be enabled by option **instrument_bound_checkers=True**.
The idea is simple,
at first we associate every buffer_var with actual Buffer shape, as far as I understood, in term of TVM buffer_var is like a pointer to actual Buffer, and then in special pass insert check before memory access.
Also new buffer could be created - we check it.
Some buffers could increase size by allocate - we check it.
Could have situation when optimization pass insert an intrinsic tvm.if_then_else - we check it.

I think it could solve the problem mentioned in the issue. 

Could someone please review and give me any feedback.
Thanks.

tqchen commented:
Thanks for the contribution, please request reviews

denis0x0D commented:
Hi @tqchen can you please review the patch.
Thanks.

denis0x0D commented:
@tqchen got it :)
Hi @tmoreau89 @ZihengJiang @yzhliu can you please review the patch.
Thanks.

ajtulloch commented:
One thing I'm curious about - have you considered instead (or in addition) adding an option for generating code for LLVM backends with address sanitizer? It seems pretty simple to add (https://github.com/halide/Halide/blob/1d471591077939fda9d9cbb77f92c6322cf28206/src/CodeGen_LLVM.cpp#L1094-L1104), and would catch these OOB accesses, along with a lot more kinds of bugs (https://clang.llvm.org/docs/AddressSanitizer.html).

denis0x0D commented:
@ajtulloch thanks for point this out.
At first I was thinking about to just add ASan's pass to backend,
but IMHO we should be really careful with it, because it could not be a generic solution:
1. ASan alongside with instrumentation pass requires runtime part, which allocates shadow memory and manages it like poison/unpoison and so on. The place of the shadow memory depends on your OS and architecture, we can face really hard bugs with it.
I personally faced some issues:
a. ASan's runtime was not initialized before instrumented binary/library - which cause a segfault.
b. GDB maps binary with PIE to ASan's shadow on 32 bit ARM Tizen OS.
c. Broken "fast unwinder" on 32 bit thumb https://gcc.gnu.org/ml/gcc-patches/2018-09/msg00312.html?fbclid=IwAR0hP_a3kFdCZbwMq07kjQnhoPdfg_3MZZya9roPYy8l8I46rpZjgj3TMM4
And so.
2. ASan officially does support 32 bit and 64 bit ARM only for Android, so other OS requires to have patched version of ASan's runtime.
3. GPU support? Depends on how it allocates memory, some issue https://github.com/google/sanitizers/issues/629

So, IMHO that every architecture which TVM supports and every new architecture would require to custom ASan's runtime, but otherwise if we need just x86/x64 it should be ok.
Thanks.

denis0x0D commented:
@ajtulloch thanks for point this out.
At first I was thinking about to just add ASan's pass to backend,
but IMHO we should be really careful with it, because it could not be a generic solution:
1. ASan alongside with instrumentation pass requires runtime part, which allocates shadow memory and manages it like poison/unpoison and so on. The place of the shadow memory depends on your OS and architecture, we can face really hard bugs with it.
I personally faced some issues:
a. ASan's runtime was not initialized before instrumented binary/library - which cause a segfault.
b. GDB maps binary with PIE to ASan's shadow on 32 bit ARM Tizen OS.
c. Broken "fast unwinder" on 32 bit thumb https://gcc.gnu.org/ml/gcc-patches/2018-09/msg00312.html?fbclid=IwAR0hP_a3kFdCZbwMq07kjQnhoPdfg_3MZZya9roPYy8l8I46rpZjgj3TMM4
And so.
2. ASan officially does support 32 bit and 64 bit ARM only for Android, so other OS requires to have patched version of ASan's runtime.
3. GPU support? Depends on how it allocates memory, some issue https://devtalk.nvidia.com/default/topic/1037466/cuda-programming-and-performance/cuda-runtime-library-and-addresssanitizer-incompatibilty/

So, IMHO that every architecture which TVM supports and every new architecture would require to custom ASan's runtime, but otherwise if we need just x86/x64 it should be ok.
Thanks.

tmoreau89 commented on python/tvm/build_module.py:
```diff
@@ -317,6 +318,10 @@ def lower(sch,
         LoopPartition, api wrapper generation and Unrolling.
 
+    instrument_bound_checkers : bool, optional
+        Whether to instruemnt loads and stores with check for out of the bound
```
"instrument" typo

denis0x0D replied:
Thanks, fixed.

tmoreau89 commented on python/tvm/build_module.py:
```diff
@@ -490,6 +499,10 @@ def build(inputs,
         By default, a new buffer is created for each tensor in the argument.
 
+    instrument_bound_checkers : bool, optional
+        Whether to instruemnt loads and stores with check for out of the bound
```
typo

denis0x0D replied:
Thanks, fixed.

tmoreau89 commented on src/pass/bound_checker.cc:
```diff
@@ -0,0 +1,135 @@
+    return index.defined() && buffer_var.defined() &&
+           BoundCheckerManager::Global()->mem_to_buffer.count(
+               buffer_var.get()) &&
+           // FIXME. Should we implement check when index is ramp ?
```
that's a good point, it may be a good idea to add support for vectorized loops

denis0x0D replied:
Thanks for mention this. I've added support for vectorized loop when index is ramp.

tmoreau89 commented:
Overall this is very good work @denis0x0D! Does it support bound checking on vectorized loops?

denis0x0D commented on python/tvm/build_module.py:
```diff
@@ -373,6 +378,9 @@ def lower(sch,
     stmt = ir_pass.RemoveNoOp(stmt)
     stmt = ir_pass.RewriteUnsafeSelect(stmt)
+    # Instrument BoundCheckers
+    if instrument_bound_checkers:
```
Thanks, fixed.

denis0x0D commented:
@tqchen @tmoreau89 thanks for review, I've updated patch regarding to your suggestions.

tmoreau89 commented:
great thank you for addressing my comments @denis0x0D!

tqchen commented:
The use of global singleton to pass information between passes seems to be dangerous, as:

- The const Variable* can get de-allocated and re-allocated to a different variable
- It makes the pass less self-contained, need to depend on context information.

I would recommend two possible ways to alleviate this:
- Do the bound checking during storage flatten
- Embed the mem_to_buffer information inside the IR itself, using AttrStmt, like https://github.com/dmlc/tvm/blob/master/include/tvm/ir.h#L208

denis0x0D commented:
@tqchen @tmoreau89 thanks for review. 
Looks like I need to learn TVM IR more precisely, I didn't really know about how  const Variable*  could be reused internally. 
The problem I've tried to solve with global hashtable - to map actual size of the buffer to buffer_var, because the actual size could be increased by other passes, which could cause a false positive. Also, I might miss this mechanism which could helps to map buffer_var to actual buffer, it might be AttrStmt.

In other case, IMHO, it seems like a normal practice to pass some context depended info between passes in other compilers, for example GCC UBSAN check for object size, at first walk on control flow graph for every basic block and insert internal call, if needed, with info like ptr to OBJECT, builtin __builtin_object_size which helps to get actual size - https://github.com/gcc-mirror/gcc/blob/master/gcc/ubsan.c#L2213
and then expand it and insert check
https://github.com/gcc-mirror/gcc/blob/master/gcc/ubsan.c#L925
Also this tool has out of bounds check, which works pretty match the same way, but the bound could not be changed between passes https://github.com/gcc-mirror/gcc/blob/master/gcc/ubsan.c#L687

denis0x0D commented:
@tqchen @tmoreau89 @yzhliu I've updated patch. Now it based on AttrStmt, could you please review it.
Thanks.

tqchen commented on src/pass/storage_flatten.cc:
```diff
@@ -96,11 +96,15 @@ class StorageFlattener : public IRMutator {
       return this->Mutate(op->body);
     } else if (op->attr_key == attr::opengl_stage_scope) {
```
can we pass create_bound_attributes as argument of storage flatten?

denis0x0D replied:
Ok, fixed, but this also required to change public api for Storage Flatten pass and some tests.

tqchen commented on src/pass/storage_flatten.cc:
```diff
@@ -440,10 +488,14 @@ class StorageFlattener : public IRMutator {
   // The current stage is an OpenGL shader.
   bool is_opengl_{false};
+  // Whether to mark load/store with theirs bounds.
+  bool create_bound_attributes{false};
```
Google C style, private member names end with underscore _

denis0x0D replied:
Thanks, fixed.

tqchen commented:
@xqdan @sgrechanik-h @ZihengJiang can you also help review this PR?

tqchen commented:
To follow up the discussion @denis0x0D 
- Variable* get reused:
    - If we transformed the IR and the original Variable* get de-allocated(because of the IR get transformed), then the transformed IR allocate another Variable*, there is a tiny chance that the de-allocated space gets allocated to the same Variable* again.
- Pass context info between passes: I agree that compiler passes should pass information to each other. The only question is that how this context info be passed, we could do
    - Global singleton context state
    - Context object
    - Embed the info in the IR

The global context object probably will cause a hard time for the programmer to reason about things because it depends on the exact order of the pass. Imagine if we do another pass in parallel and both of them write to the same global singleton context, then there would be some problem.

When possible, use an explicit context info object or embed the information into the IR will make the reasoning easier, since the pass writer only has to consider how to implement the exact signature of the function.

denis0x0D commented:
@tqchen I agree that the best way to pass info between passes is to embed it into IR, my fault was that I didn't know about AttrStmt and its semantics, and it seems like very useful node for this purpose. So, the latest version is based on AttrStmt, hope I use it in right way.
Thanks.

sgrechanik-h commented on src/api/api_pass.cc:
```diff
@@ -126,7 +126,7 @@ REGISTER_PASS1(ConvertSSA);
 REGISTER_PASS1(RewriteUnsafeSelect);
 REGISTER_PASS4(Inline);
-REGISTER_PASS3(StorageFlatten);
+REGISTER_PASS4(StorageFlatten);
```
The new parameter of StorageFlatten may be made optional, defaulting to `false`. In this case the tests calling StorageFlatten don't have to be changed. (On the other hand, making `true` the default may be a better decision in the long run, I'm not sure)

denis0x0D replied:
@sgrechanik-h thanks for review. Make sense to me to make new parameter as default.
I've executed tests locally - tests/scripts/task_python_unittest.sh and tests/scripts/task_python_topi.sh with bound checkers instrumentation by default, to verify the patch. The problem is that I'm not able to reproduce the stable tests suite same, as on remote CI, some tests fails even on master branch, and in this case I didn't find any new fails comparison to master branch, but it seems to me there are could be some tests which check the added attributes and they could fails, because if we set instrumentation to true every loads/stores instructions will mark with theirs bounds by attributes.
Thanks.

tqchen commented:
@sgrechanik-h @xqdan please check and https://docs.tvm.ai/contribute/code_review.html#approve-and-request-changes-explicitly

xqdan commented:
nice work. two questions for discussion, 
1, when we have some loop transformations in the middle, such as loop partition, the load/store has changed, can still work? could you try this case?

2, when we do this for accelerators, it is a little different. we have tensorized intrinsics and loops have been replaced by them before storage flatten. 

at least in this pr, the first question should be considered. can add a test case for it?

tmoreau89 commented:
+1 on adding a test case for accelerator back-end

@denis0x0D you can try to build a test case out of this simple tutorial example and run it with in VTA simulator (which is what is enabled by default)

denis0x0D commented:
@xqdan @tmoreau89  thanks for review. 
1. @xqdan yes it works, because instrumentation pass executes after all transformation and handles cases when buffer was reallocated and so on.
I've added more test cases for loop partitioning, splitting, tilling for simple models alongside with conv2D model. Also added some test cases for tensors with different shapes for simple op like mul.
2. @xqdan thanks for mentioning such cases,@tmoreau89 thanks for link to the tutorial.
As far as I understood even for accelerators such as VTA the "storage flatten" pass should come at first, but I might be wrong, please fix me in this case. 
VTA build_module adds passes to the first stage
https://github.com/dmlc/tvm/blob/master/vta/python/vta/build_module.py#L55
TVM build_module executes those passes after default first stage
https://github.com/dmlc/tvm/blob/master/python/tvm/build_module.py#L349

@tmoreau89 @xqdan to implement those checks for VTA, probably requires some additional logic, because current implementation requires to add attributes for every load and stores with their bounds
```
  15 produce C {
  16   for (i, 0, n) {
  17     // attr [C] buffer_bound = n
  18     // attr [B] buffer_bound = n
  19     // attr [A] buffer_bound = n
  20     if (((((int64((i + 1)) >= (int64)0) && (int64((i + 1)) < int64(n))) && ((int64(i) >= (int64)0) && (int6     4(i) < int64(n)))) && ((int64(i) >= (int64)0) && (int64(i) < int64(n))))) {
  21       C[i] = (A[(i + 1)] + B[i])
  22     } else {
  23       assert(((((int64((i + 1)) >= (int64)0) && (int64((i + 1)) < int64(n))) && ((int64(i) >= (int64)0) &&      (int64(i) < int64(n)))) && ((int64(i) >= (int64)0) && (int64(i) < int64(n)))), "OUT OF THE BOUNDS")
  24       1
  25     }
  26   }
  27 }
```
but in this case MatchCopyPattern check failes in inject_copy_intrin.cc, so probably I need to understand how it works and add another patch on top of this.
Thanks.

xqdan commented:
thanks! 
for loop partition, please make sure we can insert correct assert for tail loop body, that is the thing I wanna emphasize.

for inject copy intrin pass, the reason of failure is the if stmt inserted, we need to support assert if in this pass.

xqdan commented:
>As far as I understood even for accelerators such as VTA the "storage flatten" pass should come at first, but I might be wrong, please fix me in this case.

Tensorize is in schedule ops pass, which is before storage flatten. inject copy is another codegen pass which is after flatten, looks not hard to work with this pr. but for tensorize, we need more discussion.

denis0x0D commented on tests/python/unittest/test_pass_bound_checkers.py:
```diff
@@ -0,0 +1,399 @@
+    tvm.testing.assert_allclose(c.asnumpy(), a.asnumpy() + 1)
+
+def test_in_bounds_loop_partition_basic_llvm():
+    n = tvm.var('n')
```
@xqdan Added new test cases on const loop with check for TVM ir.

denis0x0D commented:
@xqdan 

> for loop partition, please make sure we can insert correct assert for tail loop body, that is the thing I wanna emphasize. I've added a comment for that.

I've added more tests cases. The condition for the branch should always be correct,  because it relies on  index  from load/store.

So the algorithm is straightforward - 
at first, in StorageFlatten pass we mark every load/store instruction with their bounds by AttrStmt, next
we walk recursively on created IR and collect all created AttrStmt and after that we run instrumentation pass, which inserts branch with condition for every load and store, the condition consists of two parts

- Lower bound index of the load/store should be always greater then 0, because of normalization.
- Upper bound index of the load/store should be always less then actual size of buffer associated with that load/store. The size of the buffer could be increased by Allocate node, we check that too.

Example of the condition for spliting const loop :
```
 368 produce compute {
 369   for (i.outer, 0, 5) {
 370     for (i.inner, 0, 4) {
 371       // attr [compute] buffer_bound = 21
 372       // attr [B] buffer_bound = 21
 373       // attr [A] buffer_bound = 21
 374       if (((((int64(((i.outer*4) + i.inner)) >= (int64)0) && (int64(((i.outer*4) + i.inner)) < int64(21)))      && ((int64(((i.outer*4) + i.inner)) >= (int64)0) && (int64(((i.outer*4) + i.inner)) < int64(21)))) && ((int     64(((i.outer*4) + i.inner)) >= (int64)0) && (int64(((i.outer*4) + i.inner)) < int64(21))))) {
 375         compute[((i.outer*4) + i.inner)] = (A[((i.outer*4) + i.inner)] + B[((i.outer*4) + i.inner)])
 376       } else {
 377         assert(((((int64(((i.outer*4) + i.inner)) >= (int64)0) && (int64(((i.outer*4) + i.inner)) < int64(2     1))) && ((int64(((i.outer*4) + i.inner)) >= (int64)0) && (int64(((i.outer*4) + i.inner)) < int64(21)))) &&      ((int64(((i.outer*4) + i.inner)) >= (int64)0) && (int64(((i.outer*4) + i.inner)) < int64(21)))), "OUT OF TH     E BOUNDS")
 378         1
 379       }
 380     }
 381   }
 382   for (i.inner, 0, 1) {
 383     // attr [compute] buffer_bound = 21
 384     // attr [B] buffer_bound = 21
 385     // attr [A] buffer_bound = 21
 386     if (((((int64((i.inner + 20)) >= (int64)0) && (int64((i.inner + 20)) < int64(21))) && ((int64((i.inner      + 20)) >= (int64)0) && (int64((i.inner + 20)) < int64(21)))) && ((int64((i.inner + 20)) >= (int64)0) && (in     t64((i.inner + 20)) < int64(21))))) {
 387       compute[(i.inner + 20)] = (A[(i.inner + 20)] + B[(i.inner + 20)])
 388     } else {
 389       assert(((((int64((i.inner + 20)) >= (int64)0) && (int64((i.inner + 20)) < int64(21))) && ((int64((i.i     nner + 20)) >= (int64)0) && (int64((i.inner + 20)) < int64(21)))) && ((int64((i.inner + 20)) >= (int64)0) &     & (int64((i.inner + 20)) < int64(21)))), "OUT OF THE BOUNDS")
 390       1
 391     }
 392   }
 393 }
```


> Tensorize is in scheduleOps pass, which is before storage flatten. inject copy is another codegen pass which is after flatten, looks not hard to work with this pr. but for tensorize, we need more discussion.

@xqdan  thanks for example

> for inject copy intrin pass, the reason of failure is the if stmt inserted, we need to support assert if in this pass. we can fix this issue in another PR.

Seems to me the pattern should be changed to support this kind of instrumentation, I've started working on this as part of support the instrumentations for out of the bounds for VTA, probably this should be different patch on top of this one.
Thanks.

xqdan commented:
LGTM. thanks for your work!

tqchen commented on include/tvm/ir_pass.h:
```diff
@@ -181,11 +181,11 @@ Stmt Inline(Stmt stmt,
-Stmt StorageFlatten(Stmt stmt,
-                    Map<Tensor, Buffer> extern_buffer,
-                    int cache_line_size);
+Stmt StorageFlatten(Stmt stmt, Map<Tensor, Buffer> extern_buffer,
```
nit, one line per argument

denis0x0D replied:
thanks, fixed. BTW should TVM has its own .clang-format file ?

tqchen replied:
There is a proposal in #1732 , we generally use Google C style, but haven't yet have our own .clang-format

tqchen commented on src/pass/storage_flatten.cc:
```diff
@@ -429,6 +452,30 @@ class StorageFlattener : public IRMutator {
   };
+
+  bool ShapeIsValid(const Array<Expr> &shape) {
+    if (!shape.size())
```
it is possible for us to have zero-dimensional tensors(scalar), in which case shape.size() == 0

denis0x0D replied:
Yes, I believe that it's right way to skip this type of tensors and do not instrument, please fix me if I'm wrong. 
I've added some tests cases. The compute op could be more complicated, at first I was thinking if the store accesses buffer with shape == 0, we should skip annotations and instrumentations, but there are could be example where store accesses zero shape buffer, but load accesses not zero shape.
I've added test cases for this situation. https://github.com/dmlc/tvm/pull/2079/commits/aa0745c08f90268377d9c8785208248281332f64#diff-f904c24c2b3bba4f50a565961755b153R452

tqchen replied:
I see, can you add a comment that zero-dimensional tensor does not need boundary check, so it can give context future readers.

denis0x0D replied:
Of course, added.

denis0x0D commented:
@xqdan thanks for review.

> it is possible for us to have zero-dimensional tensors, we do need to confirm if we handle this case correctly

@tqchen  thanks for review.
Yes, I believe that it's right way to skip this type of tensors and do not instrument, please fix me if I'm wrong.
I've added some test cases. The compute op could be more complicated, at first I was thinking if the store accesses buffer with shape == 0, we should skip annotations and instrumentations, but there are could be an example where store accesses zero shape buffer, but load accesses not zero shape.

```
 A = tvm.placeholder((n, ), name='A')
 scale = tvm.placeholder((), name='scale')
 k = tvm.reduce_axis((0, n), name="k")
 C = tvm.compute((), lambda : tvm.sum(A[k + k + k] * scale, axis=k), name="C")
 D = tvm.compute((), lambda : C + 1)
 s = tvm.create_schedule(D.op)

```

I've added test cases for this situation.
[first](https://github.com/dmlc/tvm/pull/2079/commits/aa0745c08f90268377d9c8785208248281332f64#diff-f904c24c2b3bba4f50a565961755b153R452)
[second](https://github.com/dmlc/tvm/pull/2079/commits/aa0745c08f90268377d9c8785208248281332f64#diff-f904c24c2b3bba4f50a565961755b153R433)

sgrechanik-h commented on src/api/api_pass.cc:
```diff
@@ -66,6 +66,11 @@ TVM_REGISTER_API("ir_pass.Equal")
+TVM_REGISTER_API("ir_pass.StorageFlatten")
+.set_body([](TVMArgs args, TVMRetValue *ret) {
+      *ret = StorageFlatten(args[0], args[1], args[2],
+                            args.size() == 3 ? false : args[3]);
```
The following would be better (to avoid duplicating the default value in two places)
```
if (args.size() <= 3) {
        *ret = StorageFlatten(args[0], args[1], args[2]);
} else {
        *ret = StorageFlatten(args[0], args[1], args[2], args[3]);
}
```

denis0x0D replied:
Thanks for review, updated.

tqchen commented on src/pass/bound_checker.cc:
```diff
@@ -0,0 +1,193 @@
+  }
+
+  bool IndexIsValid(const Expr &index) const {
+    if (!index.defined())
```
for if spans multiple lines, enclose with {}, or put return in the same line

denis0x0D replied:
thanks for review, updated.

tqchen commented:
cc @tmoreau89
